### ISSUE TITLE
`pytest.skip` (and variants) are `NoReturn` but not properly detected

### ISSUE THREAD

**Issue by mflova** [astral-sh/ty#1112]
### Summary

Maybe it is because `NoReturn` is not supported yet, but this snippet is OK for `mypy` but not for `ty`:

```py
import pytest


def my_fixture() -> (
    int
):  #  error[invalid-return-type] Function can implicitly return `None`, which is not assignable to return type `int`
    try:
        return 2
    except NotImplementedError:
        pytest.skip("Not implemented")
```

However, this is OK despite it is also based on `NoReturn`:

```py
from typing import NoReturn


def my_fixture() -> int:
    try:
        return 2
    except NotImplementedError:
        func()


def func() -> NoReturn:
    raise NotImplementedError("This function is not yet implemented")
```

Both functions are labelled as `NoReturn`. I guess it is related to the decorator attached to `skip` from `pytest`

### Version

0.0.1-alpha.19

**Comment by sharkdp** [comment]
Thank you for reporting this.

The problem is the `_with_exception` decorator on `pytest.skip`. Our generics solver is not yet capable of resolving the [`_F` typevar](https://github.com/pytest-dev/pytest/blob/8d99211f0ce3927eb7ee579f7b4f969da06dc787/src/_pytest/outcomes.py#L92-L98) in a call to `_with_exception`, and so we infer `_WithException[Unknown, <class 'Skipped'>]` for `pytest.skip`. The `Unknown` in the first type parameter means that calls to `pytest.skip` will also result in `Unknown`.

`pytest` has actually recently updated the definition of `pytest.skip`. So if you use the latest version from `pytest`s main branch, everything should work:
```py
▶ uv run --no-project --with pytest@git+https://github.com/pytest-dev/pytest ty check main.py
All checks passed!
```

Related ty issue which addresses the underlying issue (and should make this work for old pytest versions): https://github.com/astral-sh/ty/issues/623

**Comment by graipher** [comment]
@sharkdp Even with `pytest>=9`, where the original code by the OP passes, it does not work in all circumstances:

```python
from typing import reveal_type

import pytest


def test_foo(s: str | None) -> None:
    if s is None:
        pytest.skip()
    reveal_type(s)

```

```
$ uvx ty check .
info[revealed-type]: Revealed type
 --> src/foo/__init__.py:9:17
  |
7 |     if s is None:
8 |         pytest.skip()
9 |     reveal_type(s)
  |                 ^ `str | None`
  |
```

`ty 0.0.1-alpha.29`, `pytest 9.0.1`

**Comment by sharkdp** [comment]
@graipher Looks like your example is related to #685

**Comment by carljm** [comment]
@alex Out of curiosity, since you mentioned this issue is a blocker for cryptography: does it work for you on `pytest>=9`, where they changed their annotations to bypass this limitation of our current constraint solver? Or are you stuck on an earlier pytest version? Or are you still seeing symptoms like @graipher on `pytest>=9`, where the root cause problem is actually #685?

**Comment by alex** [comment]
We're on pytest 9 for python >= 3.10 -- so I suspect for my local reproduction that means its _actually_ #685. However we need to support python 3.8 and 3.9, which require older pytest, so we don't benefit from the workaround.

**Comment by fraghag** [comment]
I'm getting something similar, which I guess related to the same issue:

``` python
import pytest

pytest.fail("Test fail")
```
-> Error: Argument to bound method `fail` is incorrect: Expected `str`, found `_WithException[def fail(reason: str = "", pytrace: bool = True) -> Never, <class 'Failed'>]`

**Comment by sharkdp** [comment]
> I'm getting something similar, which I guess related to the same issue:

I can not reproduce this. Can you tell us more about your setup (ty version, pytest version)?

**Comment by fraghag** [comment]
> I can not reproduce this. Can you tell us more about your setup (ty version, pytest version)?

ty version is 0.0.15

I get it for `pytest==8.4.2`, but not `pytest==9.0.2`. They seem to have refactored that code quite a bit.
Not sure if it's pytest's or ty's "fault" in `pytest==8.4.2`.

Also: using python 3.12.5

**Comment by fraghag** [comment]
It seems like they made changes in pytest 9 to make it more spec compliant w.r.t `pytest.skip` "and friends".
https://github.com/pytest-dev/pytest/releases/tag/9.0.0 (first bullet under "Bug fixes").

So maybe this was never a an issue in ty?

Update: read through all the comments now in the thread, maybe I'm not really coming with new info here.

**Comment by carljm** [comment]
It looks to me like the originally reported issues with the `_with_exception` decorator in pytest 8 are now gone (due to changes in ty), and the new symptom is the one reported by @fraghag, which was also reported in https://github.com/astral-sh/ty/issues/2797

Closing this issue as dupe of https://github.com/astral-sh/ty/issues/2797 since that more clearly describes the current symptom, without historical baggage.

**Comment by graipher** [comment]
@carljm https://github.com/astral-sh/ty/issues/1112#issuecomment-3595005466 is still present in `ty 0.0.17`, `pytest 9.0.2`, even with https://github.com/astral-sh/ty/issues/685 already closed.

Should I open that as a new issue, since it does not seem to be related to https://github.com/astral-sh/ty/issues/2797 either?

**Comment by AlexWaygood** [comment]
> [@carljm](https://github.com/carljm) [#1112 (comment)](https://github.com/astral-sh/ty/issues/1112#issuecomment-3595005466) is still present in `ty 0.0.17`, `pytest 9.0.2`, even with [#685](https://github.com/astral-sh/ty/issues/685) already closed.

I cannot reproduce this; it appears fixed to me in 0.0.17:

```
~/dev % cat foo.py                                                  
from typing import reveal_type

import pytest


def test_foo(s: str | None) -> None:
    if s is None:
        pytest.skip()
    reveal_type(s)
~/dev % uvx --with=pytest==9.0.2 ty@0.0.16 check foo.py
info[revealed-type]: Revealed type
 --> foo.py:9:17
  |
7 |     if s is None:
8 |         pytest.skip()
9 |     reveal_type(s)
  |                 ^ `str | None`
  |

Found 1 diagnostic
~/dev % uvx --with=pytest==9.0.2 ty@0.0.17 check foo.py
info[revealed-type]: Revealed type
 --> foo.py:9:17
  |
7 |     if s is None:
8 |         pytest.skip()
9 |     reveal_type(s)
  |                 ^ `str`
  |

Found 1 diagnostic
```

**Comment by graipher** [comment]
> I cannot reproduce this; it appears fixed to me in 0.0.17:

Indeed, sorry, seems like my editor is using its own version of `ty` and that one is not quite uptodate...
